zrml/global-disputes/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+
+            let mut outcome_info =
+                <Outcomes<T>>::get(market_id, &outcome).ok_or(Error::<T>::OutcomeDoesNotExist)?;
+            let outcome_owners_len = outcome_info.owners.len() as u32;
+
+            // The `outcome_sum` never decreases (only increases) to allow
+            // caching the outcome with the highest `outcome_sum`.
+            // If the `outcome_sum` decreases, it would lead to more storage,
+            // because the winning outcome could have a smaller `outcome_sum`
+            // than the second highest `outcome_sum`.
+            let add_to_outcome_sum = |a| {
+                outcome_info.outcome_sum = outcome_info.outcome_sum.saturating_add(a);
+                <Winners<T>>::mutate(market_id, |highest| {
+                    *highest = Some(highest.clone().map_or(
+                        WinnerInfo::new(outcome.clone(), outcome_info.outcome_sum),
+                        |prev_winner_info| {
+                            if outcome_info.outcome_sum >= prev_winner_info.vote_sum {
+                                WinnerInfo::new(outcome.clone(), outcome_info.outcome_sum)
+                            } else {
+                                prev_winner_info
+                            }
+                        },
+                    ));
+                });
+                <Outcomes<T>>::insert(market_id, &outcome, outcome_info);
+            };
+
+            let vote_lock_counter = <Locks<T>>::try_mutate(
+                &sender,
+                |LockInfo(ref mut lock_info)| -> Result<u32, DispatchError> {
+                    match lock_info.binary_search_by_key(&market_id, |i| i.0) {
+                        Ok(i) => {
+                            let prev_highest_amount: BalanceOf<T> = lock_info[i].1;
+                            // concious decision here to not throw an error when the new amount is not higher
+                            // because the user would have to pay worst case fees then
+                            if amount > prev_highest_amount {
+                                let diff = amount.saturating_sub(prev_highest_amount);
+                                add_to_outcome_sum(diff);
+                                lock_info[i].1 = amount;
+                            }
+                        }
+                        Err(i) => {
+                            ensure!(
+                                lock_info.len() as u32 <= T::MaxGlobalDisputeVotes::get(),
+                                Error::<T>::MaxVotesReached
+                            );
+                            add_to_outcome_sum(amount);
+                            lock_info.insert(i, (market_id, amount));
+                        }
+                    }
+                    Ok(lock_info.len() as u32)
+                },
+            )?;
+
+            T::Currency::extend_lock(
+                T::VoteLockIdentifier::get(),
+                &sender,
+                amount,
```

Comment:
If the lock is for one global dispute / market id, then you described the behaviour right. In addition to that, you could vote on multiple outcomes for one market id / global dispute. For this you would need to divide your free balance in multiple parts. 

Say for example you have `1000 ZTG`. You throw in `300 ZTG` for outcome A, then `600 ZTG` for outcome B, then `900 ZTG` for outcome C. Then you have `900 ZTG` locked, because locks overlap. `100 ZTG` are still non-locked balance, so you are able to transfer only `100 ZTG`, [but you have still `1000 ZTG` free balance](https://github.com/zeitgeistpm/zeitgeist/blob/a88fd3e521fb850369886e6d1ea9b042b0b510af/zrml/global-disputes/src/tests.rs#L412-L451). `Locks` storages has only the highest amount saved (`900 ZTG`, so it's storage efficient, but allows also to have the ability to vote on different outcomes, so the `outcome_sum` increased for each outcome A, B and C).

If you throw in `1000 ZTG` right at the start for outcome C, you would loose the ability to vote on another outcome.

If we add up the locked amount (as in your example), we would need to keep track of each vote on each different voting outcome to prevent the unwanted behaviour, that a user can influence the final `outcome_sum` with the same funds.

But the longer I think about this, the more I come to your alluded solution, that we could add up the vote amount and compare it with the current free balance. I will try to complete this thought later.

Reply:
Thanks for the in-depth explanation. What is the reasoning for being able to vote on multiple outcomes in one dispute? I think it would make more sense if they cancel out each other, since only one of those can win in the end. I would be a ambiguous oracle (within the global disputes) if I vote for multiple outcomes.
What I wanted to point out is that across markets locks don't stack. This means I can lock 50 ZTG and use those to vote on global disputes for `MaxGlobalDisputeVotes` different markets.

Reply:
The reason to vote on multiple outcomes:
There exists one big implementation problem. This is, because we took the decision to allow infinite outcome additions. We cache the highest winner outcome. This has the drawback, that we don't like to allow the user to unlock the funds before the global dispute ends (because this would affect the final `outcome_sum`).  
Example: Outcome A in the lead with 100 ZTG, second is B with 90 ZTG, Winner info has outcome A -> user unlocks from A 20 ZTG -> B should be in the lead now, but we don't have the information about the current vote value of B, because we only cache one winner. 
So with the current solution we are storage efficient with a minimal version of the result, we'd like to have. The drawback is, that the users funds are not liquid until the global dispute period is over.

If we'd only allow to vote on one outcome in one dispute, then we'd need to save the outcome index (which is not the problem) AND we would restrict the user to vote again only for the same outcome. When we allow multiple outcomes, then we don't need to store the information about the last outcome (which was voted on) and the user has still the flexibility to add additional voting funds to another outcome, if the user made some mistake with the first call.

Across different markets (/ global disputes) the locks overlap, which is another flexibility I would like to see, so that the user can lock the same funds for `pallet_democracy` and has the maximum possible voice on different global disputes.

Reply:
In most cases, the behaviour of the user is: Voting on the believed correct outcome only ones.

Reply:
I wonder if we rather do this to stack the voting amount. This would lead to the expected behaviour.

```rust
match lock_info.binary_search_by_key(&market_id, |i| i.0) {
    Ok(i) => {
        let prev_highest_amount: BalanceOf<T> = lock_info[i].1;
        let new_highest_amount = amount.saturating_add(prev_highest_amount);
        ensure!(new_highest_amount <= T::Currency::free_balance(&voter), Error::<T>::InsufficientAmount);
        let diff = amount.saturating_sub(prev_highest_amount);
        add_to_outcome_sum(diff);
        lock_info[i].1 = new_highest_amount;
    }
    Err(i) => {
        lock_info
            .try_insert(i, (market_id, amount))
            .map_err(|_| Error::<T>::MaxVotesReached)?;
        add_to_outcome_sum(amount);
    }
}
```

Example: You have 1000 ZTG.
You vote with 300 ZTG on outcome A, then 300 ZTG on outcome B, then 300 ZTG on outcome C. Now you would have locked 900 ZTG (`new_highest_amount`).

I think this is correct. Could you please double check?

Reply:
ed85132844a47350bb5392de2a59eeecd3a04fcc